tests/drivers/counter/counter_basic_api/src/test_counter.c

```diff
@@ -801,7 +842,24 @@ void test_cancelled_alarm_does_not_expire(void)
 
 void test_main(void)
 {
+	struct device *dev;
+	int i;
+
+	k_sem_init(&top_cnt_sem, 0, UINT_MAX);
+	k_object_access_grant(&top_cnt_sem, k_current_get());
+
+	k_sem_init(&alarm_cnt_sem, 0, UINT_MAX);
+	k_object_access_grant(&alarm_cnt_sem, k_current_get());
+
+	for (i = 0; i < ARRAY_SIZE(devices); i++) {
+		dev = device_get_binding(devices[i]);
+		zassert_not_null(dev, "Unable to get counter device %s",
+				 devices[i]);
+		k_object_access_grant(dev, k_current_get());
+	}
+
 	ztest_test_suite(test_counter,
+		/* Uses callbacks, run in supervisor mode */
```

Comment:
you would get more code coverage if you set up the callbacks in the driver in a supervisor thread, and then perform the remainder of the test actions in user mode.

see tests/drivers/uart/uart_async_api/src/main.c for an example

Reply:
I will look into that.

Reply:
That is cleverly done. I would like to get more test coverage of the user mode handlers, but using a similar approach requires restructuring of the existing tests to separate setup and execution.

Since the missing syscalls are blocking #21846 (which I would like to get in before 2.2) I suggest merging this reduced test case for user mode along with the syscalls and then address the refactoring and increase of test coverage at a later point.